crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -212,3 +284,119 @@ pub enum Error {
     #[error(transparent)]
     InvalidExecutedAmount(#[from] InvalidExecutedAmount),
 }
+
+// todo: should be removed once integration tests are implemented
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_adjusted_price_improvement_amounts_for_sell_order() {
+        let order_sell_amount = to_wei(20);
+        let order_buy_amount = to_wei(19);
+        let quote_sell_amount = to_wei(21);
+        let quote_buy_amount = to_wei(18);
+        let quote_fee_amount = to_wei(1);
+
+        let (sell_amount, _) = adjusted_price_improvement_amounts(
+            order_sell_amount,
+            order_buy_amount,
+            Side::Sell,
+            quote_sell_amount,
+            quote_buy_amount,
+            quote_fee_amount,
+        )
+        .unwrap();
+
+        assert_eq!(
+            sell_amount, order_sell_amount,
+            "Sell amount should match order sell amount for sell orders."
+        );
+    }
+
+    #[test]
+    fn test_adjusted_price_improvement_amounts_for_buy_order() {
+        let order_sell_amount = to_wei(20);
+        let order_buy_amount = to_wei(19);
+        let quote_sell_amount = to_wei(21);
+        let quote_buy_amount = to_wei(18);
+        let quote_fee_amount = to_wei(1);
+
+        let (_, buy_amount) = adjusted_price_improvement_amounts(
+            order_sell_amount,
+            order_buy_amount,
+            Side::Buy,
+            quote_sell_amount,
+            quote_buy_amount,
+            quote_fee_amount,
+        )
+        .unwrap();
+
+        assert_eq!(
+            buy_amount, order_buy_amount,
+            "Buy amount should match order buy amount for buy orders."
+        );
+    }
+
+    #[test]
+    fn test_sell_order_with_quote_in_market_price() {
+        let order_sell_amount = to_wei(10);
+        let order_buy_amount = to_wei(20);
+        let quote_sell_amount = to_wei(9);
+        let quote_buy_amount = to_wei(25);
+        let quote_fee_amount = to_wei(1);
+
+        let (sell_amount, buy_amount) = adjusted_price_improvement_amounts(
+            order_sell_amount,
+            order_buy_amount,
+            Side::Sell,
+            quote_sell_amount,
+            quote_buy_amount,
+            quote_fee_amount,
+        )
+        .unwrap();
+
+        assert_eq!(
+            sell_amount, order_sell_amount,
+            "Sell amount should be taken from the quote for sell orders in market price."
```

Comment:
With this comment I would expect the assertion to be `sell_amount == quote_sell_amount`.